Parser percent arrays fix

MSP-Greg opened the pull request:
# Description

Fixes percent array parsing with `\n ]` for closing delimiter
# Completed Tasks
- [X] I have read the [Contributing Guide](https://github.com/lsegal/yard/blob/master/CONTRIBUTING.md).
- [X] The pull request is complete (implemented / written).
- [X] Git commits have been cleaned up (squash WIP / revert commits).
- [X] I wrote tests and they pass (if code is attached to PR).

lsegal commented on spec/parser/ruby/ruby_parser_spec.rb:
```diff
@@ -283,6 +283,54 @@ def hello; end
       end
     end
 
+    it "correcly parses several formats of percent arrays in constant declarations" do
```
You can rewrite this single test as a series of `it()` blocks (move the `it` call into the `str_a.each` block). That way you don't have to do your own failure output management at the bottom-- rspec can handle this automatically.

MSP-Greg commented:
Thanks for the explanation of some options.

Personally, I would prefer to have the error message formatted as is, since, by listing on consecutive lines, it's very clear where the parsing error is.  For instance, what if in the future Ripper separates the 'left side delimiter' from the %i, %I, %w , or %W characters?  The failure output would make it very clear.

I'll admit to having a bias towards tabular data.  I think one comment thread I alluded to that regarding log messages...

Hence, what would you like me to do?

lsegal commented:
I would recommend making the `it()` change. The extra granularity on the tests allows for better understanding of which tests failed, and allows us to run individual tests rather than all of them together (very useful when debugging a specific parsing issue with logging). If you really want the output text, you can always add the `"#{str_sub.ljust(31)} ->  #{p_source}"` data to the msg argument of the individual `expect()` calls. Best of both worlds.

lsegal commented on spec/parser/ruby/ruby_parser_spec.rb:
```diff
@@ -283,6 +283,54 @@ def hello; end
+        ['|', '|']
+      ]
+
+      type_a = [
```
This may as well be a hash for readability (since order doesn't matter for 1.8 anyway)

MSP-Greg commented:
First, I changed both the arrays of two element arrays to hashes. then I changed the block structure, then commented out the one line in ruby_parser.  Ran RSpec, and honestly, I thought it was a mess.

We've got a 5 (delimiters) x 4 (%i, %I, etc) x 5 (whitespace) set of tests, and chances are, one item in one dimension/vector may fail in the future.  So we'll probably have at least 20 or 25 failures.  Then again, could be more, or could be less.  Without the tabular format, it is almost impossible to collate the info into anything meaningful.

Hence, I left it as is, but I did update the description to make clear all the tests it performs.  We'll hope it 
doesn't fail for a while.

Tested, robo'd, commit rebased and pushed...

lsegal commented:
> We've got a 5 (delimiters) x 4 (%i, %I, etc) x 5 (whitespace) set of tests, and chances are, one item in one dimension/vector may fail in the future.

Why would these tests fail in the future? If we're expecting them to fail at some point, then they should be re-written to be less brittle. If you're suggesting that a significant change in the parser may make all of them fail, I have no problem with this.

Ultimately this is not really acceptable in the current form. `it` descriptions by convention should not contain paragraph-long descriptions or newlines. Such a long it description really proves that these are multiple independent tests. I would accept this patch if the test is split up as recommended.

lsegal commented on spec/parser/ruby/ruby_parser_spec.rb:
```diff
@@ -283,6 +283,58 @@ def hello; end
+      }
+
+      str_a = [
+        'TEST = %{type}[A B C]',
```
Testing each delimiter type seems unnecessary. There is no separate code path for each delimiter in YARD's codebase, so this is a detail of the underlying library and doesn't need testing. In other words, testing for a single delimiter should be sufficient-- if a separate delimiter fails, that would not be a bug with this patch (there is no logic for delimiters in the parser).

lsegal commented on spec/parser/ruby/ruby_parser_spec.rb:
```diff
@@ -283,6 +283,58 @@ def hello; end
+         "     in constant declarations.  Loops thru all combinations of\n" \
+         "     () {} [] <> || delimiters, %i %I %w %W array types, and five\n" \
+         "     types of element/delimiter whitespace." do
+      delimiter_h = {
```
Testing each delimiter is unnecessary. There is no codepath for delimiters in any of the parser code. Testing with a single delimiter should be sufficient-- if `%w()` worked and `%w{}` failed, that would not be due to a bug in YARD's codebase.

MSP-Greg commented:
@lsegal I'll revise this.  Been busy with some other things.

> If you're suggesting that a significant change in the parser may make all of them fail, I have no problem with this.

Certainly not all of them.  I'm suggesting, having worked on the percent array bug and the unary parameter bug, that Ripper is designed more for one-way parsing, rather than returning info for source code 'dice and chop'.  Hence, somewhat brittle for its application in YARD.

I'll revise my work files and create a repo for them, and keep the full tests in there.  If in the future, changes in Ripper result in some bugs creeping back in, at least the full tests will still be available.  Of course, they may not be helpful...

lsegal commented:
Merged as c9c9c0aca8bd9d4dbc39775855c4e129e70b108f with tests fixes at c27a05b8ea73fbb9aebb67fe4f238a2203cafc05

MSP-Greg commented:
@lsegal Sorry and thanks, been busy with Ruby & Windows...

Since you might be rev'ing YARD, when using Ruby 2.4/2.5, there's a few places  in `lib/yard/parser/ruby/legacy/ruby_lex.rb` where TRUE / FALSE generates depreciation warning.

Lines 643, 645, and 1167, 1168.  That's the only issues that I've seen running under 2.5.0...

lsegal commented:
Patches would be accepted for that, but the ruby_lex is 1.8 specific legacy code so I probably won't get around to this without a PR.

MSP-Greg commented:
I'll throw one together, not tonite.  The warnings happen on load/require, and I recall there was somewhere where a method of the class was used.  I think you even had a 'todo' type comment around the code (or the call).  Wasn't a tag though.

I may be confusing it with the legacy parser.  Maybe I'll try to find that code.
